File: database/src/model/story.py
from db.db_entity import DBEntity


class Story(DBEntity):
    def __init__(self, title, year, theme, tags, synopsis):
        super().__init__('Libros', {'title': title})

        self.title = title
        self.year = year
        self.theme = theme
        self.tags = tags
        self.synopsis = synopsis
        self.writters = []
        self.illustrators = []
        self.editorial = None

    def to_dict(self):
        return {
            'titulo': self.title,
            'anio': self.year,
            'tema': self.theme,
            'tags': self.tags,
            'sinopsis': self.synopsis,
            'escritores': self.writters,
            'ilustradores': self.illustrators,
            'editorial': self.editorial
        }

    def add_writter(self, writter):
        self.writters.append(writter)

    def add_illustrator(self, illustrator):
        self.illustrators.append(illustrator)

    def set_editorial(self, editorial):
        self.editorial = editorial

    def __repr__(self):
        return 'Story(title={}, year={}, theme={}, tags={}, synopsis={})'.format(self.title,
                                                                                 self.year,
                                                                                 self.theme,
                                                                                 self.tags,
                                                                                 self.synopsis)

File: visualizations/app/story.js
const _db = Symbol('db');

export default class Story {
  constructor(db, qs) {
    this[_db] = db;

    const data = qs.data();

    this['titulo'] = data['titulo'];
    this['anio'] = data['anio'];
    this['tema'] = data['tema'];
    this['sinopsis'] = data['sinopsis'];
    this['tags'] = data['tags'];

    this['editorial'] = data['editorial'];
    this['ilustradores'] = data['ilustradores'];
    this['escritores'] = data['escritores'];
  }

  fetch_editorial() {
    return this.fetchAttribute('editorial');
  }

  fetch_illustrators() {
    return this.fetchAttributes('ilustradores');
  }

  fetch_writters() {
    return this.fetchAttributes('escritores');
  }

  fetchAttribute(attr) {
    return this[attr].get()
      .then(value => {
        this[attr] = value.data();
        return this;
      });
  }

  fetchAttributes(attr) {
    return Promise.all(this[attr].map(a => a.get()))
      .then(values => {
        this[attr] = values.map(v => v.data());
        return this;
      });
  }

  detail() {
    return Promise.all(
      [
        this['editorial'].get(),
        this['editors'].get(),
        ...this['illustrators'].map(i => i.get()),
        ...this['writters'].map(w => w.get())
      ])
      .then(values => {
        values = values.map(v => v.data());

        this['editorial'] = values[0];
        this['editor'] = values[1];
        this['illustrators'] = values[2];
        this['writters'] = values[3];

        return this;
      })
  }
}
